src/SBSpergel.cpp

```diff
@@ -0,0 +1,715 @@
+#endif
+        }
+
+    }
+
+    void SBSpergel::SBSpergelImpl::fillKValue(tmv::MatrixView<std::complex<double> > val,
+                                            double x0, double dx, double dxy,
+                                            double y0, double dy, double dyx) const
+    {
+        dbg<<"SBSpergel fillKValue\n";
+        dbg<<"x = "<<x0<<" + ix * "<<dx<<" + iy * "<<dxy<<std::endl;
+        dbg<<"y = "<<y0<<" + ix * "<<dyx<<" + iy * "<<dy<<std::endl;
+        assert(val.stepi() == 1);
+        assert(val.canLinearize());
+        const int m = val.colsize();
+        const int n = val.rowsize();
+        typedef tmv::VIt<std::complex<double>,1,tmv::NonConj> It;
+
+        x0 *= _r0;
+        dx *= _r0;
+        dxy *= _r0;
+        y0 *= _r0;
+        dy *= _r0;
+        dyx *= _r0;
+
+        It valit = val.linearView().begin();
+        for (int j=0;j<n;++j,x0+=dxy,y0+=dy) {
+            double x = x0;
+            double y = y0;
+            for (int i=0;i<m;++i,x+=dx,y+=dyx) {
+                double ksq = x*x + y*y;
+                *valit++ = _flux * _info->kValue(ksq);
+            }
+        }
+    }
+
+    SpergelInfo::SpergelInfo(double nu, double trunc, const GSParamsPtr& gsparams) :
+        _nu(nu), _trunc(trunc), _gsparams(gsparams),
+        _gamma_nup1(boost::math::tgamma(_nu+1.0)),
+        _gamma_nup2(_gamma_nup1 * (_nu+1)),
+        _xnorm0((_nu > 0.) ? _gamma_nup1 / (2. * _nu) * std::pow(2., _nu) : INFINITY),
+        _truncated(_trunc > 0.),
+        _maxk(0.), _stepk(0.), _re(0.), _flux(0.),
+        _ft(Table<double,double>::spline)
+    {
+        dbg<<"Start SpergelInfo constructor for nu = "<<_nu<<std::endl;
+        dbg<<"trunc = "<<_trunc<<std::endl;
+
+        if (_nu < sbp::minimum_spergel_nu || _nu > sbp::maximum_spergel_nu)
+            throw SBError("Requested Spergel index out of range");
+    }
+
+    class SpergelIntegratedFlux
+    {
+    public:
+        SpergelIntegratedFlux(double nu, double gamma_nup2, double flux_frac=0.0)
+            : _nu(nu), _gamma_nup2(gamma_nup2),  _target(flux_frac) {}
+
+        double operator()(double u) const
+        // Return flux integrated up to radius `u` in units of r0, minus `flux_frac`
+        // (i.e., make a residual so this can be used to search for a target flux.
+        {
+            double fnup1 = std::pow(u / 2., _nu+1.)
+                * boost::math::cyl_bessel_k(_nu+1., u)
+                / _gamma_nup2;
+            double f = 1.0 - 2.0 * (1.+_nu)*fnup1;
+            return f - _target;
+        }
+    private:
+        double _nu;
+        double _gamma_nup2;
+        double _target;
+    };
+
+    double SpergelInfo::calculateFluxRadius(const double& flux_frac) const
+    {
+        // Calcute r such that L(r/r0) / L_tot == flux_frac
+
+        // These seem to bracket pretty much every reasonable possibility
+        // that I checked in Mathematica, though I'm not very happy about
+        // the lower bound....
```

Comment:
I'm more concerned about the upper bound.  The input `flux_frac` could be arbitrarily close to 1 if the user puts in a very small `folding_threshold` or `shoot_accuracy`.  Are you sure that 20 is always safe for that?

I think maybe if flux_frac < 0.5, do the call to `bracketLowerWithLimit` as you have here to make sure the lower bound is ok, otherwise call `bracketUpper` to make sure the upper bound is ok.